groups/ntc/ntcq/ntcq_zerocopy.cpp

```diff
@@ -19,187 +19,451 @@
+    BSLS_ASSERT(d_waitList.empty() || d_waitList.front().group() < group);
 
-        if (callback) {
-            ntca::SendEvent sendEvent;
-            sendEvent.setType(eventType);
-            sendEvent.setContext(d_sendContext);
+    ntcq::ZeroCopyCounter counter = d_generator.next();
 
-            callback
-                .dispatch(sender, sendEvent, strand, executor, defer, mutex);
-        }
-        ++it;
+    bsl::shared_ptr<ntsa::Data> dataContainer =
+        d_dataPool_sp->createOutgoingData();
+
+    *dataContainer = data;
+
+    d_waitList.resize(d_waitList.size() + 1);
+    ZeroCopyEntry& entry = d_waitList.back();
+
+    entry.setGroup(group);
+    entry.setMinCounter(counter);
+    entry.setMaxCounter(counter + 1);
+    entry.setData(dataContainer);
+
+    if (callback) {
+        entry.setCallback(callback);
     }
-    d_callbacks.clear();
+
+    return counter;
 }
 
-ZeroCopyWaitList::ZeroCopyWaitList(bslma::Allocator* basicAllocator)
-: d_entryList(basicAllocator)
-, d_strand()
-, d_nextId(0)
+ntcq::ZeroCopyCounter ZeroCopyQueue::push(
+    ntcq::SendCounter                  group, 
+    const bsl::shared_ptr<ntsa::Data>& data)
 {
+    BSLS_ASSERT(d_waitList.empty() || d_waitList.front().group() < group);
+
+    ntcq::ZeroCopyCounter counter = d_generator.next();
+
+    d_waitList.resize(d_waitList.size() + 1);
+    ZeroCopyEntry& entry = d_waitList.back();
+
+    entry.setGroup(group);
+    entry.setMinCounter(counter);
+    entry.setMaxCounter(counter + 1);
+    entry.setData(data);
+
+    return counter;
 }
 
-ZeroCopyWaitList::~ZeroCopyWaitList()
+ntcq::ZeroCopyCounter ZeroCopyQueue::push(
+    ntcq::SendCounter                  group,
+    const bsl::shared_ptr<ntsa::Data>& data, 
+    const ntci::SendCallback&          callback)
 {
-    BSLS_ASSERT(d_entryList.empty());
+    BSLS_ASSERT(d_waitList.empty() || d_waitList.front().group() < group);
+
+    ntcq::ZeroCopyCounter counter = d_generator.next();
+
+    d_waitList.resize(d_waitList.size() + 1);
+    ZeroCopyEntry& entry = d_waitList.back();
+
+    entry.setGroup(group);
+    entry.setMinCounter(counter);
+    entry.setMaxCounter(counter + 1);
+    entry.setData(data);
+
+    if (callback) {
+        entry.setCallback(callback);
+    }
+
+    return counter;
 }
 
-void ZeroCopyWaitList::setStrand(const bsl::shared_ptr<ntci::Strand>& strand)
+ntcq::ZeroCopyCounter ZeroCopyQueue::push(ntcq::SendCounter group)
 {
-    d_strand = strand;
+    NTCCFG_WARNING_UNUSED(group);
+
+    ntcq::ZeroCopyCounter counter = d_generator.next();
+
+    ZeroCopyEntry& entry = d_waitList.back();
+    BSLS_ASSERT(entry.group() == group);
+
+    entry.setMaxCounter(counter + 1);
+
+    return counter;
 }
 
-void ZeroCopyWaitList::addEntry(ZeroCopyEntry& entry)
+void ZeroCopyQueue::frame(ntcq::SendCounter group)
 {
-    BSLS_ASSERT(!cancelled);
+    NTCCFG_WARNING_UNUSED(group);
 
-    entry.setId(d_nextId++);
+    ZeroCopyEntry& entry = d_waitList.back();
```

Comment:
Do we rely on assertion built into ```bsl::list``` or we should add our on assertion on top?
```BSLS_ASSERT(!d_waitList.empty());```
I think it can happen that for some reason BDE is build w/o enabled asserts.

Reply:
bsl::list is a template that is built with our definition of asserts. That being said, we assert on the required state of `d_waitList` in the other overloads of `ntcq::ZeroCopyQueue::push()`; I'll add an assert to this function as well, for consistency. Done.

Reply:
So do you plan to add an asertion here or I missed something?

Reply:
I added the assert.